packages/sol/src/availabilitySlots.ts

```diff
@@ -0,0 +1,144 @@
+  const filtered = mainSubslots.filter(sub => !isIntersecting(sub, bookedSubslots));
+  
+  // Check if the booking slot is a subslot of at least one of the filtered list;
+  // if so return true, otherwise return false
+  const bookingSlot: IAvailabilitySlot.SubSlot = {
+    parent: -1,
+    start: bookInfo.start,
+    end: dayjs(bookInfo.start).add(bookInfo.duration, "minutes").toISOString(),
+  };
+  for (const slot of filtered) {
+    if (isSuperSlot(slot, bookingSlot))
+      return true;
+  }
+  return false;
+}
+
+/**
+ * removes the booked slots (represented by subslots) from the `slots` and 
+ * then returns the free remaining space/time represented by subslots. 
+ */
+export function subtractSlots({
+  slots,
+  subslots,
+}: {
+   slots: IAvailabilitySlot.Slot[];
+   subslots: IAvailabilitySlot.SubSlot[];
+}): IAvailabilitySlot.SubSlot[] {
+  // get the smallest subslot duration
+  let smallestDur = min(subslots.map(sub => dayjs(sub.end).diff(sub.start, "minutes")));
+  if (!smallestDur) throw Error("evaluating duration failed; check the passed subslots.");
+  
+  // divide the slots using the smallest evaluated duration
+  const mainSubslots = getSubslotsBatch(slots, smallestDur);
+
+  // remove from the (divided) subslots the ones that intersects
+  // with at least one (passed parameter) subslot.
+  return mainSubslots.filter(sub => !isIntersecting(sub, subslots));
+}
+
+// checks if a specific slot intersects with at least one slot of the passed list.
+// the boundaries are included in the intersection.
+export function isIntersecting(
+  target: IAvailabilitySlot.GeneralSlot, 
+  slots: IAvailabilitySlot.GeneralSlot[]
+): boolean {
+  for (const slot of slots) {
+    const startTimeCondition = dayjs.utc(target.start).isAfter(slot.end) || 
+      dayjs.utc(target.start).isSame(slot.end);
+    if (startTimeCondition) continue;
+
+    const endTimeCondition = dayjs.utc(target.end).isBefore(slot.start) || 
+      dayjs.utc(target.end).isSame(slot.start);
+    if (endTimeCondition) continue;
+
+    return true;
+  }
+  return false;
+}
+
+/**
+ *  checks if slot `a` can be considered as a super slot (contains) for slot `b` or not.
+ */
+export function isSuperSlot(
+  a: IAvailabilitySlot.GeneralSlot, 
+  b: IAvailabilitySlot.GeneralSlot
+): boolean {
+  if (dayjs.utc(a.start).isAfter(b.start)) 
+    return false;
+  if (dayjs.utc(a.end).isBefore(b.end)) 
+    return false;
+  return true;
+}
+
+// sort a list of slots/subslots. note: return a new copy and don't mutate the input. 
+export function sortSlots(
+  slots: IAvailabilitySlot.GeneralSlot[], 
+  dir: 'asc' | 'desc'
+): IAvailabilitySlot.GeneralSlot[] {
+  const cloned = clone(slots);
+  return cloned.sort((a, b) => {
+    const dirValue = dir === "asc" ? 1 : -1;
+    if (dayjs.utc(a.start).isBefore(b.start))
+      return -1 * dirValue;
+    else if (dayjs.utc(a.start).isAfter(b.start))
+      return 1 * dirValue
+    else
+      return 0;
+  });
```

Comment:
The function is hard to understand from the first look. Let's use lodash `orderBy` for convenience. It is used [here](https://github.com/litespace-org/litespace/blob/ddc09149172b24c29ab774f769854ede9c66860a/services/server/src/lib/tutor.
ts#L105).